alfredo/train/ppo/train.py

```diff
@@ -0,0 +1,391 @@
+# Copyright 2023 The Brax Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Proximal policy optimization training.
+
+See: https://arxiv.org/pdf/1707.06347.pdf
+"""
+
+import functools
+import time
+from typing import Callable, Optional, Tuple, Union
+
+from absl import logging
+from brax import envs
+from brax.training import acting
+from brax.training import gradients
+from brax.training import pmap
+from brax.training import types
+from brax.training.acme import running_statistics
+from brax.training.acme import specs
+from brax.training.agents.ppo import losses as ppo_losses
+from brax.training.agents.ppo import networks as ppo_networks
+from brax.training.types import Params
+from brax.training.types import PRNGKey
+from brax.v1 import envs as envs_v1
+import flax
+import jax
+import jax.numpy as jnp
+import optax
+
+InferenceParams = Tuple[running_statistics.NestedMeanStd, Params]
+Metrics = types.Metrics
+
+_PMAP_AXIS_NAME = 'i'
+
+
+@flax.struct.dataclass
+class TrainingState:
+  """Contains training state for the learner."""
+  optimizer_state: optax.OptState
+  params: ppo_losses.PPONetworkParams
+  normalizer_params: running_statistics.RunningStatisticsState
+  env_steps: jnp.ndarray
+
+
+def _unpmap(v):
```

Comment:
If we could import this functions that are available in brax we could shrink this file

Reply:
We could import _unpmap() and _strip_weak_type() from Brax - any other function you were thinking of?